applications/OptimizationApplication/python_scripts/controls/material/simp_control.py

```diff
@@ -0,0 +1,199 @@
+
+    def Check(self) -> None:
+        self.filter.Check()
+
+    def Finalize(self) -> None:
+        self.filter.Finalize()
+
+    def GetPhysicalKratosVariables(self) -> 'list[SupportedSensitivityFieldVariableTypes]':
+        return self.controlled_physical_variables
+
+    def GetEmptyField(self) -> ContainerExpressionTypes:
+        field = Kratos.Expression.ElementExpression(self.model_part)
+        Kratos.Expression.LiteralExpressionIO.SetData(field, 0.0)
+        return field
+
+    def GetControlField(self) -> ContainerExpressionTypes:
+        return self.phi.Clone()
+
+    def MapGradient(self, physical_gradient_variable_container_expression_map: 'dict[SupportedSensitivityFieldVariableTypes, ContainerExpressionTypes]') -> ContainerExpressionTypes:
+        keys = physical_gradient_variable_container_expression_map.keys()
+        if len(keys) != 2:
+            raise RuntimeError(f"Not provided required gradient fields for control \"{self.GetName()}\". Following are the variables:\n\t" + "\n\t".join([k.Name() for k in keys]))
+        if  Kratos.DENSITY not in keys or Kratos.YOUNG_MODULUS not in keys:
+            raise RuntimeError(f"The required gradient for control \"{self.GetName()}\" w.r.t. DENSITY or YOUNG_MODULUS not found. Followings are the variables:\n\t" + "\n\t".join([k.Name() for k in keys]))
+
+        # first calculate the density partial sensitivity of the response function
+        d_j_d_density = physical_gradient_variable_container_expression_map[Kratos.DENSITY]
+
+        # second calculate the young modulus partial sensitivity of the response function
+        d_j_d_youngs = physical_gradient_variable_container_expression_map[Kratos.YOUNG_MODULUS]
+
+        # now calculate the total sensitivities of density w.r.t. phi
+        d_density_d_phi = KratosOA.ControlUtils.SigmoidalProjectionUtils.CalculateForwardProjectionGradient(self.phi, self.materials.GetPhi(), self.materials.GetDensities(), self.beta, 1)
+
+        # now calculate the total sensitivities of young modulus w.r.t. phi
+        d_young_modulus_d_phi = KratosOA.ControlUtils.SigmoidalProjectionUtils.CalculateForwardProjectionGradient(self.phi, self.materials.GetPhi(), self.materials.GetYoungModulus(), self.beta, self.simp_power_fac)
+
+        # now compute response function total sensitivity w.r.t. phi
+        d_j_d_phi = d_j_d_density * d_density_d_phi + d_j_d_youngs * d_young_modulus_d_phi
+
+        return self.filter.FilterIntegratedField(d_j_d_phi)
+
+    def Update(self, control_field: ContainerExpressionTypes) -> bool:
+        if not IsSameContainerExpression(control_field, self.GetEmptyField()):
+            raise RuntimeError(f"Updates for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.model_part.FullName()}, given model part name: {control_field.GetModelPart().FullName()} ]")
+
+        if Kratos.Expression.Utils.NormL2(self.phi - control_field) > 1e-15:
```

Comment:
here is mistaken. Since phi is control field we are checking if it is updated or not and then we have to reset phi wi th the new control field and then filter phi/new control field and then project. but currently phi is reset with the filtered phi which is a mistake